### DISCUSSION TITLE
Difference Between vpm UPDATE and UPGRADE?

### DISCUSSION THREAD

**Discussion by omegaui** [vlang/v#18177] [Questions and Answers] [answered]
Hello!
I am currently designing a **full fledged V Lang IDE.**
the repo is private until stable version lands.

The IDE will **support each and every tip of V's functionalities**,
I am currently working on managing modules though IDE dialog.

And I want to know the difference between `vpm update` and `vpm upgrade`.

This is how the IDE's launcher view looks right now.
![Screenshot from 2023-05-14 21-24-01](https://github.com/vlang/v/assets/73544069/a1b9aa71-95f4-4652-a8ff-0210a957daf7)

and This is the Module Management View I am designing,
![image](https://github.com/vlang/v/assets/73544069/83b1025c-55c6-4850-8cba-8ed05412e102)

And this is the design to be implemented for which I need the answer.

![module-manager-light](https://github.com/vlang/v/assets/73544069/a71a30b3-65c1-49e0-8ea4-9bd0789496a6)

**Bonus: The IDE already supports auto installing both V and git on all of the three desktop platforms through its first start up Setup Tour**.

**Comment by JalonSolov** [comment]
```
$ v help update
Updates each PACKAGE.

Usage:
  v update [PACKAGE]...

With no PACKAGE, ALL installed packages are updated to their latest versions.

Options:
  -v           Print more details about the update.
  -server-url  When doing network operations, use this vpm server.
               Can be given multiple times.

$ v help upgrade
Upgrade all outdated modules.

Usage:
  v upgrade

Options:
  -v  Print more details about the upgrade.

$
```

**Reply by omegaui** [reply]
@JalonSolov 

This is something I ain't getting!

If `update` can check and update all the packages or the provided to list to the latest version (from their outdated version to latest ones),
then, what does `upgrade` doing actually by saying the word **upgrading** outdated versions.

btw, Thank you for helping me on this!

**Reply by omegaui** [reply]
What It feels to me is that both update and upgrade are doing the same thing :thinking:

**Reply by omegaui** [reply]
And are `packages` and `modules` different?

**Reply by omegaui** [reply]
I mean I have publish a module my self, but what is a `package` here?

**Reply by JalonSolov** [reply, marked as answer]
`package` is incorrect here.  It _should_ say `module`.  Not sure why it hasn't been corrected.

The difference between the 2 is that `upgrade` will update _all_ modules you have installed, whether you want them to be updated or not (and some people like to stick to specific releases of modules, for whatever reason).

`update` without specifying a module does do the same as `upgrade`, but if you do specify a module, it will only do that single one and ignore the rest.

So, yes, `upgrade` probably could go away, but... it doesn't hurt anything to have both.

**Reply by omegaui** [reply]
Thanks!!
You just cleared my confusion.

**Reply by Wajinn** [reply]
Just adding some experiences that I've had.  Using `update` has always worked well for me, but using `upgrade` has caused problems in the past, where I had to reinstall modules.  It's easy enough to reinstall modules, so never looked into it, but just putting that out there.

I do like the flexibility of being able to update specific modules.  So as pointed out, having both doesn't seem to be a problem, as long as the documentation clearly explains the difference.

**Comment by omegaui** [comment]
And I'm talking specifically about `vpm` and not `v`.
